Documentation/articles/intro.md

```diff
@@ -1,5 +1,182 @@
+overhead to pay simply to use the code, but once you are up and running it is possible to
+process messages with no per-message allocations.
+
+The benchmarks built into the library source repository include a test that extracts location
+information from a file containing 1 million messages, and this has exactly the same memory
+allocation characteristics as a test that reads 1,000, or 10 million messages, demonstrating
+that after the first use, we make zero allocations per-message.
+
+Ais.Net achieves this by taking advantage of `Span<T>` and related types which were added in
+.NET Core 2.1, and using associated language features added in C# 7.2. This makes Ais.Net
+somewhat different to use from other .NET AIS parsers.
+
+The `Span<T>` and `ReadOnlySpan<T>` types impose a constraint in exchange for the high
+performance they make possible: these types can only live on the stack. This prevents variables
+of these types from living in fields of other types unless those containing types are
+defined in the same way. (So if you define a type that has the same constraint—if you define
+a `ref struct` type, it too can only live on the stack—it is allowed to contain other
+`ref struct` types such as `Span<T>`.) For this reason the various payload parsers that
+Ais.Net defines, such as
+[NmeaAisPositionReportClassAParser](https://ais-dotnet.github.io/Ais.Net/api/Ais.Net.NmeaAisPositionReportClassAParser.html),
+can also only live on the stack. So you can do this:
+
+```csharp
+public static (double Latitude, double Longitude) GetPosition(byte[] asciiNmeaLine)
+{
+    var parser = new NmeaAisPositionReportClassAParser(asciiNmeaLine);
+    return (parser.Latitude10000thMins / 600000, parser.Longitude10000thMins / 600000);
+}
+```
+
+But this will not work:
+
+```csharp
+public class PositionExtractor
+{
+    // This will produce a compiler error, because you cannot have a ref struct
+    // a field in a non-ref-struct type.
+    private readonly NmeaAisPositionReportClassAParser parser;
+    public PositionExtractor(byte[] asciiNmeaLine)
+    {
+        this.parser = new NmeaAisPositionReportClassAParser(asciiNmeaLine);
+    }
+}
+```
+
+More subtly, this restriction prevents you from using `ref struct` types as local variable
+in `async` methods. To enable `async` methods to continue after an `await` completes
+asynchronously, C# needs to be able to store all local variables on the heap, which means
+those variables must not include any `ref struct` types. This means you cannot use the various
+Ais.Net `Parser` types as variables in such methods. If you need to parse an NMEA message from
+an asynchronous method, you can use a nested method, e.g.:
+
+```csharp
+public static async Task<(double latitude, double longitude)> GetPositionAsync()
+{
+    ReadOnlyMemory<byte> line = await GetLineAsync();
+    return ParseLine();
+
+    (double latitude, double longitude) ParseLine()
```

Comment:
This is a very nice technique that deserves broader dissemination.

Reply:
Blog draft in progress...